New Instrument View - Rectangle selection on 2D projection

GuiMacielPereira opened the pull request:
### Description of work
Adds rectangular picking of multiple detectors on both 2D and 3D in a consist manner.

<!-- Please provide an outline and reasoning for the work.
If there is no linked issue provide context.
-->

Closes [39886](https://github.com/mantidproject/mantid/issues/39886). <!-- One line per closed issue. -->

<!-- If issue raised by user. Do not leak email addresses.
**Report to:** [user name]
-->

### To test:
Tick and untick checkbox for multiple selection of detectors and try different combinations of both 3D and 2D projections.


<!-- Include sufficient instructions for someone unfamiliar with the application to test.
Ok to refer back to instructions in the issue.
-->

<!-- REMEMBER:
- Add labels, milestones, etc.
- Ensure the base of this PR is correct (e.g. release-next or main)
- Add release notes in separate file as per ([guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)), or justify their absence:
  *This does not require release notes* because <fill in an explanation of why>
-->

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

GuiMacielPereira committed:
Add file for 2D and 3D interactor styles

GuiMacielPereira committed:
Add interactor styles to enable rectangle picking in 2d

These interactor styles allow for rectangular picking with right button
on both 3D and 2D projections.

On the 2D projection case, this is achieved by layering two vtk styles
on top of the interactor, and disabling certain observers from both
styles by overwriting them to do nothing.

As a word of caution, calling `removeObserver` or `removeAllObservers`
on a vtk style does not remove default observers, the only way that I
found around this is to instead overwrite default observers using
`AddObserver`.

GuiMacielPereira committed:
Fix issues with changing projection with checkbox ticked

If multi picking checkbox is ticked when we change projection, re-update
view as if checkbox was clicked.

GuiMacielPereira committed:
Update unit tests

GuiMacielPereira committed:
Clean code from CodeRabbit suggestions

jclarkeSTFC reviewed (changes requested):
Works great, it's good to have a consistent behaviour across the 2D and 3D views. I think that `_ReversedInteractorStyleRubberBand2D.on_selection_changed` could do with a unit test, maybe one for checking `_clamp_to_window` works as well.

RichardWaiteSTFC reviewed (changes requested):
I know I've not been asked to look at this so sorry for sticking my nose in!
But the 2D rectangular selection is quite slow if you're not summing the spectra - probably because you're plotting so many lines on the 1D plot (possibly hundreds!). For single-crystal at least for this use case will always want to sum spectra - perhaps it would be good to disable the checkbox when rectangular selection is used?

GuiMacielPereira committed:
Add unit tests for interactor styles

GuiMacielPereira committed:
Disable rectangle selecting if not sum

GuiMacielPereira committed:
Do rect select always in sum

- Remove previous enabling/disabling rectangular select checkbox
- Multiple select is always plotted in sum of spectra

GuiMacielPereira committed:
Set sum checkbox checked if rect select clicked

RichardWaiteSTFC reviewed (dismissed):
Looks great, thanks for changes to sum spectra checkbox!

jclarkeSTFC commented:
I think there's a genuine problem on macOS that needs investigating, so I disabled auto-merge

sf1919 commented:
SHould this be pointed at main instead?

GuiMacielPereira committed:
Simplify unit tests to avoid library issues

jclarkeSTFC reviewed (approved):
Tests fixed by mocking out calls to `OpenGL`, which can cause problems on virtual machines.
